### PR TITLE
Replace custom code with grpcurl go library for reflection calls

### PR DESCRIPTION
**Pull request by nela-cc**
### Description
- Replaced custom reflection code with grpcul go lib.
- For simplicity, all endpoint files were removed, and the logic was moved to client.go. Removed all the reflection response redefinition.
- Slightly updated the tests.

### Parent Issue
Issue #215 

### Comments for the Reviewers
- Test with make eccs-tests

### Type(s) of Change (Split the PR if you check many)
- [ ] Added user feature
- [ ] Added technical feature
- [ ] Added tests
- [x] Refactor
- [ ] Bugfix
- [ ] Updated documentation
- [ ] Infrastructure changes (Terraform, GCP, K8s, other)
- [ ] CI/CD workflow changes

### Testing Checklist
- [x] I have added/updated unit tests for functional changes.
- [ ] I have added/updated integration tests for changes that affect dependent systems.
- [ ] I have added/updated end-to-end tests for feature changes.

### General Checklist
- [ ] I have pulled in the latest changes from master.
- [x] I have run `make lint`.
- [x] I have successfully run `make tests`.
- [ ] I have updated relevant workflows and deployment tools.
- [ ] I have updated/added relevant documentation (readme, doc comments, etc).
- [ ] I have added the license to new source code files.
- [x] I have spent some time looking over the full diff before creating this PR.

### CODE REVIEW

File: applications/ECCS/app/client.go [hunk 1]
```diff
@@ -83,571 +123,413 @@ func NewClient(userAT string) (*Client, error) {
...
-		}
+	err = grpcurl.InvokeRPC(
+		c.ctx,
+		c.reflSource,
+		c.connection,
+		method,
+		c.authHeader,
+		handler,
+		rf.Next)
+	if err != nil {
```
Comment (philip-cc) [inline applications/ECCS/app/client.go:177]: Something is wrong with the error handling here. If the user is not authorized, the call succeeds but there is not data returned.

File: applications/ECCS/cmd/root.go [hunk 2]
```diff
@@ -60,127 +60,155 @@ var storeCmd = &cobra.Command{
 	Use:   "store",
 	Short: "Stores your secrets using Encryptonize",
 	RunE: func(cmd *cobra.Command, args []string) error {
-		err := app.Store(userAT, filename, associatedData, stdin)
+		client, err := app.NewClient(userAT)
```
Comment (philip-cc) [inline applications/ECCS/cmd/root.go:63]: We could just do this once by creating a global variable in `InitCmd`.

File: applications/ECCS/app/client.go (hunk 1 above)
Comment (nela-cc) [reply applications/ECCS/app/client.go:178]: Added status code check to fail if request was rejected.

File: applications/ECCS/cmd/root.go (hunk 2 above)
Comment (nela-cc) [reply applications/ECCS/cmd/root.go]: Added client initialization to PreRunE function.

File: applications/ECCS/cmd/root.go [hunk 3]
```diff
@@ -59,131 +62,146 @@ Environment Variables:
 var storeCmd = &cobra.Command{
 	Use:   "store",
 	Short: "Stores your secrets using Encryptonize",
+	PreRunE: func(cmd *cobra.Command, args []string) error {
+		return InitClient(cmd, args)
+	},
```
Comment (philip-cc) [inline applications/ECCS/cmd/root.go:67]: You can just do this:
```suggestion
	PreRunE: InitClient,
```

Comment (nela-cc) [reply applications/ECCS/cmd/root.go]: Done.

File: applications/ECCS/scripts/function_tests.py [hunk 4]
```diff
@@ -202,14 +197,23 @@ def decrypt_object(token, filename):
...
 	update_object(at1, oid, new_obj["data"], new_obj["associated_data"])
 
 	d, ad = retrieve_object(at1, oid)
-	print(f"[+] object was updated successfully: {d == new_obj['data'] and ad == new_obj['associated_data']}")
+	if(d != new_obj['data'] or ad != new_obj['associated_data']):
+		print(f"[-] failed to update object")
+		sys.exit(1)
 
 	subprocess.run(["./eccs", "-a", at1, "store", "-f", "README.md", "-d", "asdf"], check=True)
 	subprocess.run(["./eccs", "-a", at1, "retrieve", "-o", oid], check=True)
```
Comment (philip-cc) [inline applications/ECCS/scripts/function_tests.py:219]: Aren't these already tested above?

File: applications/ECCS/cmd/root.go [hunk 5]
```diff
@@ -57,133 +60,124 @@ Environment Variables:
...
+	Short:   "Stores your secrets using Encryptonize",
+	PreRunE: InitClient,
 	RunE: func(cmd *cobra.Command, args []string) error {
-		err := app.Store(userAT, filename, associatedData, stdin)
-		if err != nil {
-			return err
-		}
-		return nil
+		return client.Store(filename, associatedData, stdin)
 	},
```
[the remark this answers is not in the document]
Comment (nela-cc) [reply applications/ECCS/cmd/root.go:68]: Done in all commands.

File: applications/ECCS/app/client.go [hunk 6]
```diff
@@ -83,571 +124,416 @@ func NewClient(userAT string) (*Client, error) {
...
+		IncludeTextSeparator:  true,
+		AllowUnknownFields:    false,
 	}
 
-	// resolve the method of the service
-	methDes := srvDes.FindMethodByName(method)
-	if methDes == nil {
-		errMsg := fmt.Sprintf("Service %s has no method named %s", service, method)
-		return nil, errors.New(errMsg)
+	rf, formatter, err := grpcurl.RequestParserAndFormatter(grpcurl.Format("json"), c.reflSource, in, options)
```
[the remark this answers is not in the document]
Comment (nela-cc) [reply applications/ECCS/app/client.go:158]: Done.

File: applications/ECCS/app/client.go [hunk 7]
```diff
@@ -83,571 +124,416 @@ func NewClient(userAT string) (*Client, error) {
...
+	log.Printf("%v\n%s", utils.Pass("Successfully decrypted object!"), response)
 
-	return objID, ciphertext, aad, nil
+	return nil
 }
 
-func (c *Client) Decrypt(oid string, ciphertext []byte, associatedData []byte) ([]byte, []byte, error) {
-	mth, err := c.findMethod("enc.Encryptonize", "Decrypt")
+// openFile loads a file into memory
+func openFile(filename string) []byte {
```
[the remark this answers is not in the document]
Comment (nela-cc) [reply applications/ECCS/app/client.go:510]: Deleted function, placed in readInput.

File: applications/ECCS/scripts/function_tests.py (hunk 4 above)
Comment (nela-cc) [reply applications/ECCS/scripts/function_tests.py:219]: Those are indeed already rested above, removed.

File: applications/ECCS/app/client.go [hunk 8]
```diff
@@ -41,14 +43,53 @@ type Client struct {
...
+
+type EncryptedData struct {
+	Ciphertext     string `json:"ciphertext"`
+	AssociatedData string `json:"associatedData"`
+	ObjectID       string `json:"objectId"`
+}
+
+type RetrievedData struct {
+	Plaintext      []byte `json:"plaintext"`
+	AssociatedData []byte `json:"associatedData"` //TODO: change to associated_data (encryptonize core api) and replace with Data
```
[the remark this answers is not in the document]
Comment (nela-cc) [reply applications/ECCS/app/client.go:72]: As discussed, the inconsistency is a result from input/output format difference. The change no longer applicable.

File: applications/ECCS/app/client.go [hunk 9]
```diff
@@ -83,571 +124,416 @@ func NewClient(userAT string) (*Client, error) {
...
+	object, err := json.Marshal(
+		struct {
+			ObjectID       string `json:"object_id"`
+			Plaintext      []byte `json:"plaintext"`
+			AssociatedData []byte `json:"associated_data"`
+		}{
+			ObjectID:       oid,
+			Plaintext:      plaintext,
+			AssociatedData: []byte(associatedData),
+		})
```
[the remark this answers is not in the document]
Comment (nela-cc) [reply applications/ECCS/app/client.go:254]: Extracted all to a separate file.

**Review by philip-cc — APPROVED** [review]
Ship it!

**Review by philip-cc — APPROVED** [review]

**Pull request merged**